File: xtask/tests/architecture_docs.rs
//! Current architecture documentation coherence tests.

#![forbid(unsafe_code)]

use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, Output};

use serde_json::{json, Value};

fn workspace_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("xtask lives under the workspace root")
        .to_path_buf()
}

fn run_checker(root: &Path) -> Output {
    Command::new("python3")
        .arg(workspace_root().join("scripts/architecture_docs.py"))
        .arg(root)
        .arg("--check")
        .output()
        .expect("architecture docs checker must launch")
}

fn write_json(path: &Path, value: &Value) {
    fs::write(path, serde_json::to_vec_pretty(value).unwrap()).unwrap();
}

fn current_header(title: &str, body: &str) -> String {
    format!("# {title}\n\nLast verified: 2026-08-04\n\nVyre 0.7.9.\n\n{body}\n")
}

fn write_fixture(root: &Path) {
    for directory in ["docs/generated", "docs/rfcs", "release/evidence/backends"] {
        fs::create_dir_all(root.join(directory)).unwrap();
    }
    fs::write(
        root.join("Cargo.toml"),
        "[workspace]\nresolver = \"2\"\nmembers = [\"a\", \"vyre-megakernel\"]\n",
    )
    .unwrap();
    fs::write(
        root.join("release/release-train.toml"),
        "[versions]\nvyre = \"0.7.9\"\n",
    )
    .unwrap();
    fs::write(
        root.join("docs/CRATE_OWNERSHIP.toml"),
        "schema_version = 1\n\n[[crate]]\npackage = \"vyre-megakernel\"\npath = \"vyre-megakernel\"\nowner = \"megakernel-compiler\"\nlayer = \"compiler-boundary\"\nresponsibility = \"Compile validated ProgramGraph inputs into canonical static and persistent megakernel artifacts without owning admission, execution, or lifecycle policy.\"\nallowed_dependencies = [\"vyre-foundation\"]\n",
    )
    .unwrap();
    write_json(
        &root.join("docs/generated/OP_SCHEMA.json"),
        &json!({
            "schema_version": 2,
            "operation_count": 4,
            "tier_counts": {"intrinsic": 1, "primitive": 1, "libs": 1, "runtime": 1},
            "operations": [
                {"id": "i", "tier": "intrinsic"},
                {"id": "p", "tier": "primitive"},
                {"id": "l", "tier": "libs"},
                {"id": "r", "tier": "runtime"}
            ]
        }),
    );
    write_json(
        &root.join("release/evidence/backends/backend-matrix.json"),
        &json!({
            "preferred_backend_id": "cuda",
            "blockers": [],
            "backends": [
                {"id": "cuda"},
                {"id": "wgpu"}
            ]
        }),
    );

    fs::write(
        root.join("docs/ARCHITECTURE.md"),
        current_header(
            "Architecture",
            "Use generated/OP_SCHEMA.json. The semantic authority is vyre-foundation::operation_registry; derived catalogs do not own shadow operation identities. Evidence selects CUDA as the preferred backend. Cross-program composition target. vyre-megakernel emits Artifact envelopes. Persistent protocol lives in vyre-runtime/src/megakernel/. The older bytecode interpreter design is superseded.",
        ),
    )
    .unwrap();
    fs::write(
        root.join("docs/OPTIMIZATION_ARCHITECTURE.md"),
        current_header(
            "Optimization architecture",
            "Layer 1: semantic IR optimization. Layer 2: concrete lowering strategy. Current scheduling is in vyre-runtime/src/megakernel/. Artifact freeze is owned by vyre-megakernel. IR pre-dispatch fusion lives in vyre-foundation/src/optimizer/megakernel.",
        ),
    )
    .unwrap();
    fs::write(
        root.join("docs/RUNTIME_PIPELINE.md"),
        current_header(
            "Runtime pipeline",
            "Cache: vyre-runtime/src/pipeline_cache/. Runtime: vyre-runtime/src/megakernel/. Neutral artifacts come from vyre-megakernel and enter through artifact_admission. Failure does not silently rerun. Prose does not substitute for raw samples.",
        ),
    )
    .unwrap();
    fs::write(
        root.join("docs/megakernel-wiring.md"),
        current_header(
            "Megakernel wiring",
            "Execution starts from the same validated `Program` and does not consume a general VIR bytecode interpreter. Ownership: vyre-runtime/src/megakernel/. Artifact compiler vyre-megakernel emits Artifact. Wave policy lives in vyre-driver/src/megakernel_execution. IR fusion lives in vyre-foundation/src/optimizer/megakernel. Residue note: vyre-driver-wgpu/src/megakernel was removed.",
        ),
    )
    .unwrap();
    fs::write(
        root.join("docs/rfcs/0005-persistent-megakernel.md"),
        "# RFC\n\nLast verified: 2026-08-04\n\nStatus: **Superseded**\n\nHistorical motivation. Superseded design. Current resolution. Vyre does not support a general interpreter. `vyre-megakernel` is a current workspace member that compiles typed graphs to canonical artifacts.\n",
    )
    .unwrap();
    fs::write(
        root.join("docs/DOCS.toml"),
        "schema_version = 1\n\n[[page]]\npath = \"ARCHITECTURE.md\"\nstatus = \"current\"\n\n[[page]]\npath = \"OPTIMIZATION_ARCHITECTURE.md\"\nstatus = \"current\"\n\n[[page]]\npath = \"RUNTIME_PIPELINE.md\"\nstatus = \"current\"\n\n[[page]]\npath = \"megakernel-wiring.md\"\nstatus = \"current\"\n\n[[page]]\npath = \"rfcs/0005-persistent-megakernel.md\"\nstatus = \"superseded\"\n",
    )
    .unwrap();
}

/// Current architecture guides must agree with live workspace, operation, backend, and ownership authorities.
#[test]
fn workspace_architecture_documents_are_current() {
    let output = run_checker(&workspace_root());
    assert!(
        output.status.success(),
        "Fix: repair current architecture docs: {}",
        String::from_utf8_lossy(&output.stderr)
    );
}

/// A complete fixture proves the checker accepts the current ownership split.
#[test]
fn coherent_architecture_fixture_passes() {
    let temp = tempfile::tempdir().unwrap();
    write_fixture(temp.path());
    assert!(
        run_checker(temp.path()).status.success(),
        "Fix: coherent architecture fixture must pass: {}",
        String::from_utf8_lossy(&run_checker(temp.path()).stderr)
    );
}

/// A retired release claim must not return to a current architecture guide.
#[test]
fn retired_architecture_version_fails_closed() {
    let temp = tempfile::tempdir().unwrap();
    write_fixture(temp.path());
    let path = temp.path().join("docs/RUNTIME_PIPELINE.md");
    fs::write(
        &path,
        format!("{}\nVyre 0.6.9.\n", fs::read_to_string(&path).unwrap()),
    )
    .unwrap();
    let output = run_checker(temp.path());
    assert!(!output.status.success());
    assert!(String::from_utf8_lossy(&output.stderr).contains("stale architecture pattern"));
}

/// WGPU cannot be presented as the primary production route when evidence selects CUDA.
#[test]
fn wgpu_primary_route_fails_closed() {
    let temp = tempfile::tempdir().unwrap();
    write_fixture(temp.path());
    let path = temp.path().join("docs/ARCHITECTURE.md");
    fs::write(
        &path,
        format!(
            "{}\nWGPU is the primary production path.\n",
            fs::read_to_string(&path).unwrap()
        ),
    )
    .unwrap();
    let output = run_checker(temp.path());
    assert!(!output.status.success());
    assert!(String::from_utf8_lossy(&output.stderr).contains("primary production path"));
}

/// Docs must not revive the pre-crate planned/absent claim for vyre-megakernel.
#[test]
fn planned_absent_megakernel_claim_fails_closed() {
    let temp = tempfile::tempdir().unwrap();
    write_fixture(temp.path());
    let path = temp.path().join("docs/ARCHITECTURE.md");
    fs::write(
        &path,
        format!(
            "{}\nThe planned `vyre-megakernel` crate is not a current workspace member.\n",
            fs::read_to_string(&path).unwrap()
        ),
    )
    .unwrap();
    let output = run_checker(temp.path());
    assert!(!output.status.success());
    let stderr = String::from_utf8_lossy(&output.stderr);
    assert!(
        stderr.contains("stale architecture pattern")
            || stderr.contains("architecture still presents"),
        "Fix: planned/absent claim must fail closed; stderr={stderr}"
    );
}

/// Removing the semantic registry authority must fail the joined operation model.
#[test]
fn missing_semantic_operation_registry_fails_closed() {
    let temp = tempfile::tempdir().unwrap();
    write_fixture(temp.path());
    let path = temp.path().join("docs/ARCHITECTURE.md");
    let text = fs::read_to_string(&path)
        .unwrap()
        .replace("vyre-foundation::operation_registry", "runtime registry");
    fs::write(&path, text).unwrap();
    let output = run_checker(temp.path());
    assert!(!output.status.success());
    assert!(String::from_utf8_lossy(&output.stderr).contains("operation_registry"));
}

/// A preferred backend without an executable row must not support an architecture claim.
#[test]
fn missing_preferred_backend_probe_fails_closed() {
    let temp = tempfile::tempdir().unwrap();
    write_fixture(temp.path());
    let path = temp
        .path()
        .join("release/evidence/backends/backend-matrix.json");
    let mut matrix: Value = serde_json::from_str(&fs::read_to_string(&path).unwrap()).unwrap();
    matrix["backends"] = json!([{"id": "wgpu"}]);
    write_json(&path, &matrix);
    let output = run_checker(temp.path());
    assert!(!output.status.success());
    assert!(String::from_utf8_lossy(&output.stderr).contains("no executable probe row"));
}

/// The bytecode-interpreter RFC must remain visibly superseded in both file and manifest.
#[test]
fn megakernel_rfc_cannot_return_to_current_status() {
    let temp = tempfile::tempdir().unwrap();
    write_fixture(temp.path());
    let path = temp.path().join("docs/DOCS.toml");
    fs::write(
        &path,
        fs::read_to_string(&path).unwrap().replace(
            "path = \"rfcs/0005-persistent-megakernel.md\"\nstatus = \"superseded\"",
            "path = \"rfcs/0005-persistent-megakernel.md\"\nstatus = \"current\"",
        ),
    )
    .unwrap();
    let output = run_checker(temp.path());
    assert!(!output.status.success());
    assert!(String::from_utf8_lossy(&output.stderr).contains("classify"));
}

/// A current architecture page must remain current in the documentation manifest.
#[test]
fn architecture_page_manifest_status_fails_closed() {
    let temp = tempfile::tempdir().unwrap();
    write_fixture(temp.path());
    let path = temp.path().join("docs/DOCS.toml");
    fs::write(
        &path,
        fs::read_to_string(&path).unwrap().replace(
            "path = \"ARCHITECTURE.md\"\nstatus = \"current\"",
            "path = \"ARCHITECTURE.md\"\nstatus = \"archived\"",
        ),
    )
    .unwrap();
    let output = run_checker(temp.path());
    assert!(!output.status.success());
    assert!(String::from_utf8_lossy(&output.stderr).contains("ARCHITECTURE.md"));
}

File: scripts/architecture_docs.py
#!/usr/bin/env python3
"""Validate current architecture documentation against live repository authorities."""

from __future__ import annotations

import argparse
import json
import re
import sys
import tomllib
from pathlib import Path
from typing import Any

MAX_INPUT_BYTES = 16_777_216
CURRENT_DOCS = [
    Path("docs/ARCHITECTURE.md"),
    Path("docs/OPTIMIZATION_ARCHITECTURE.md"),
    Path("docs/RUNTIME_PIPELINE.md"),
    Path("docs/megakernel-wiring.md"),
]
RFC = Path("docs/rfcs/0005-persistent-megakernel.md")
MANIFEST = Path("docs/DOCS.toml")


# Wire version of docs/generated/OP_SCHEMA.json. xtask owns the generator
# (xtask/src/operation_schema.rs SCHEMA_VERSION) and pins the same number;
# its test the_python_contract_pins_the_same_operation_schema_version fails
# when this line drifts from it.
OPERATION_SCHEMA_VERSION = 4


class ContractError(Exception):
    """Architecture prose disagrees with a live authority."""


def read_text(path: Path) -> str:
    try:
        size = path.stat().st_size
    except OSError as error:
        raise ContractError(f"could not inspect `{path}`: {error}") from error
    if size > MAX_INPUT_BYTES:
        raise ContractError(f"`{path}` exceeds the {MAX_INPUT_BYTES}-byte input limit")
    try:
        return path.read_text(encoding="utf-8")
    except (OSError, UnicodeError) as error:
        raise ContractError(f"could not read `{path}`: {error}") from error


def read_toml(path: Path) -> dict[str, Any]:
    try:
        return tomllib.loads(read_text(path))
    except tomllib.TOMLDecodeError as error:
        raise ContractError(f"could not parse TOML `{path}`: {error}") from error


def read_json(path: Path) -> dict[str, Any]:
    try:
        value = json.loads(read_text(path))
    except json.JSONDecodeError as error:
        raise ContractError(f"could not parse JSON `{path}`: {error}") from error
    if not isinstance(value, dict):
        raise ContractError(f"`{path}` must contain an object")
    return value


def normalize(text: str) -> str:
    return " ".join(text.split())


def require_tokens(path: Path, text: str, tokens: list[str]) -> None:
    normalized_text = normalize(text)
    for token in tokens:
        if normalize(token) not in normalized_text:
            raise ContractError(f"`{path}` is missing current architecture token `{token}`")


def forbid_patterns(path: Path, text: str, patterns: list[str]) -> None:
    for pattern in patterns:
        if re.search(pattern, text, re.IGNORECASE):
            raise ContractError(f"`{path}` retains stale architecture pattern `{pattern}`")


def verification_date(path: Path, text: str) -> str:
    match = re.search(r"^Last verified: (\d{4}-\d{2}-\d{2})$", text, re.MULTILINE)
    if match is None:
        raise ContractError(f"`{path}` has no Last verified date")
    return match.group(1)


def crate_rows(ownership: dict[str, Any]) -> list[dict[str, Any]]:
    rows = ownership.get("crate")
    if not isinstance(rows, list):
        raise ContractError("ownership registry has no [[crate]] table")
    out: list[dict[str, Any]] = []
    for row in rows:
        if isinstance(row, dict):
            out.append(row)
    if not out:
        raise ContractError("ownership registry has no crate rows")
    return out


def validate(root: Path) -> None:
    root = root.resolve()
    workspace = read_toml(root / "Cargo.toml")
    members = workspace.get("workspace", {}).get("members")
    if not isinstance(members, list) or not members:
        raise ContractError("workspace.members must be a non-empty array")
    if any(not isinstance(member, str) or "*" in member for member in members):
        raise ContractError("workspace members must be explicit paths")
    if "vyre-megakernel" not in members:
        raise ContractError("workspace.members must include vyre-megakernel")

    train = read_toml(root / "release/release-train.toml")
    version = train.get("versions", {}).get("vyre")
    if not isinstance(version, str):
        raise ContractError("release train has no versions.vyre")

    operation_schema = read_json(root / "docs/generated/OP_SCHEMA.json")
    operations = operation_schema.get("operations")
    tier_counts = operation_schema.get("tier_counts")
    if (
        operation_schema.get("schema_version") != OPERATION_SCHEMA_VERSION
        or not isinstance(operations, list)
        or not isinstance(tier_counts, dict)
        or operation_schema.get("operation_count") != len(operations)
        or sum(tier_counts.values()) != len(operations)
    ):
        raise ContractError("operation schema is not internally coherent")

    backend = read_json(root / "release/evidence/backends/backend-matrix.json")
    if backend.get("blockers") != []:
        raise ContractError(f"backend evidence has blockers: {backend.get('blockers')}")
    preferred = backend.get("preferred_backend_id")
    backend_rows = backend.get("backends")
    if not isinstance(preferred, str) or not isinstance(backend_rows, list):
        raise ContractError("backend evidence has no preferred backend or probe rows")
    if preferred not in {row.get("id") for row in backend_rows if isinstance(row, dict)}:
        raise ContractError("preferred backend has no executable probe row")

    ownership = read_toml(root / "docs/CRATE_OWNERSHIP.toml")
    megakernel_rows = [
        row
        for row in crate_rows(ownership)
        if row.get("package") == "vyre-megakernel" or row.get("path") == "vyre-megakernel"
    ]
    if not megakernel_rows:
        raise ContractError("ownership registry has no current vyre-megakernel crate row")
    compiler_responsibility = megakernel_rows[0].get("responsibility")
    if not isinstance(compiler_responsibility, str) or "ProgramGraph" not in compiler_responsibility:
        raise ContractError("vyre-megakernel compiler responsibility is incomplete")
    if ownership.get("planned", {}).get("vyre-megakernel") is not None:
        raise ContractError("ownership registry must not keep planned.vyre-megakernel after the crate exists")

    docs_manifest = read_toml(root / MANIFEST)
    page_rows = docs_manifest.get("page")
    if not isinstance(page_rows, list):
        raise ContractError(f"`{MANIFEST}` has no [[page]] rows")
    page_status = {
        Path("docs") / row["path"]: row.get("status")
        for row in page_rows
        if isinstance(row, dict) and isinstance(row.get("path"), str)
    }
    texts: dict[Path, str] = {}
    for path in CURRENT_DOCS:
        text = read_text(root / path)
        texts[path] = text
        verification_date(path, text)
        if version not in text:
            raise ContractError(f"`{path}` is not verified against Vyre {version}")
        if page_status.get(path) != "current":
            raise ContractError(f"`{MANIFEST}` must classify `{path}` as current")

    rfc_text = read_text(root / RFC)
    texts[RFC] = rfc_text
    verification_date(RFC, rfc_text)
    if "Status: **Superseded**" not in rfc_text:
        raise ContractError(f"`{RFC}` must be explicitly superseded")
    if page_status.get(RFC) != "superseded":
        raise ContractError(f"`{MANIFEST}` must classify `{RFC}` as superseded")

    stale_absent = [
        r"planned\s+`?vyre-megakernel`?",
        r"planned compiler crate",
        r"not a current workspace",
        r"not present in the current",
        r"Until that crate exists",
        r"Until that boundary ships",
        r"declared target rather than a shipped package",
    ]
    for path, text in texts.items():
        forbidden = [
            r"\b0\.6(?:\.x)?\b",
            r"\b9[- ]op\b",
            r"\bnine[- ]op\b",
            r"WGPU[^\n]{0,40}primary production path",
            r"## Four CI laws",
            r"codex-[0-9a-z]+",
        ]
        forbid_patterns(path, text, forbidden)
        if path in CURRENT_DOCS or path == RFC:
            forbid_patterns(path, text, stale_absent)

    architecture = texts[Path("docs/ARCHITECTURE.md")]
    require_tokens(
        Path("docs/ARCHITECTURE.md"),
        architecture,
        [
            "generated/OP_SCHEMA.json",
            "vyre-foundation::operation::OperationRegistry",
            "do not own shadow operation identities",
            "Cross-program composition",
            "vyre-megakernel",
            "Artifact",
            "vyre-runtime",
            "bytecode interpreter",
        ],
    )

    require_tokens(
        Path("docs/OPTIMIZATION_ARCHITECTURE.md"),
        texts[Path("docs/OPTIMIZATION_ARCHITECTURE.md")],
        [
            "Layer 1: semantic IR optimization",
            "Layer 2: concrete lowering strategy",
            "vyre-runtime/src/megakernel/",
            "vyre-megakernel",
            "vyre-foundation/src/optimizer/megakernel",
        ],
    )
    require_tokens(
        Path("docs/RUNTIME_PIPELINE.md"),
        texts[Path("docs/RUNTIME_PIPELINE.md")],
        [
            "vyre-runtime/src/pipeline_cache/",
            "vyre-runtime/src/megakernel/",
            "vyre-megakernel",
            "artifact_admission",
            "does not silently rerun",
            "not substitute for raw samples",
        ],
    )
    require_tokens(
        Path("docs/megakernel-wiring.md"),
        texts[Path("docs/megakernel-wiring.md")],
        [
            "starts from the same validated `Program`",
            "vyre-runtime/src/megakernel/",
            "vyre-megakernel",
            "Artifact",
            "vyre-driver/src/megakernel_execution",
            "vyre-foundation/src/optimizer/megakernel",
            "does not consume a general VIR bytecode interpreter",
        ],
    )
    require_tokens(
        RFC,
        rfc_text,
        [
            "Historical motivation",
            "Superseded design",
            "Current resolution",
            "does not support a general",
            "vyre-megakernel",
            "workspace member",
        ],
    )


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("root", type=Path, help="workspace root to validate")
    parser.add_argument(
        "--check",
        action="store_true",
        help="validate only; exit non-zero on contract failure",
    )
    args = parser.parse_args()
    try:
        validate(args.root)
    except ContractError as error:
        print(f"architecture docs contract failed: {error}", file=sys.stderr)
        return 1
    if args.check:
        print("architecture docs contract ok")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
